src/backend/distributed/planner/multi_planner.c

```diff
@@ -122,38 +124,152 @@ multi_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
-		 * using the extended outfuncs infrastructure, but it's infeasible to
-		 * test most plans. MultiQueryContainerNode always serializes the
-		 * physical plan, so there's no need to check that separately.
+		 * Router didn't yield a plan, try the full distributed planner. As
+		 * real-time/task-tracker don't support prepared statement parameters,
+		 * skip planning in that case (we'll later trigger an error in that
+		 * case if necessary).
 		 */
-		CheckNodeIsDumpable((Node *) logicalPlan);
+		if ((!distributedPlan || distributedPlan->planningError) && !hasUnresolvedParams)
+		{
+			/* Create and optimize logical plan */
+			MultiTreeRoot *logicalPlan = MultiLogicalPlanCreate(query);
+			MultiLogicalPlanOptimize(logicalPlan);
+
+			/*
+			 * This check is here to make it likely that all node types used in
+			 * Citus are dumpable. Explain can dump logical and physical plans
+			 * using the extended outfuncs infrastructure, but it's infeasible to
+			 * test most plans. MultiQueryContainerNode always serializes the
+			 * physical plan, so there's no need to check that separately.
+			 */
+			CheckNodeIsDumpable((Node *) logicalPlan);
+
+			/* Create the physical plan */
+			distributedPlan = MultiPhysicalPlanCreate(logicalPlan);
+
+			/* distributed plan currently should always succeed or error out */
+			Assert(distributedPlan && distributedPlan->planningError == NULL);
+		}
+	}
 
-		/* Create the physical plan */
-		physicalPlan = MultiPhysicalPlanCreate(logicalPlan);
+	/*
+	 * If no plan was generated, prepare a generic error to be emitted.
+	 * Normally this error message will never returned to the user, as it's
+	 * usually due to unresolved prepared statement parameters - in that case
+	 * the logic below will force a custom plan (i.e. with parameters bound to
+	 * specific values) to be generated.  But sql (not plpgsql) functions
+	 * unfortunately don't go through a codepath supporting custom plans - so
+	 * we still need to have an error prepared.
+	 */
+	if (!distributedPlan)
+	{
+		/* currently always should have a more specific error otherwise */
+		Assert(hasUnresolvedParams);
+		distributedPlan = CitusMakeNode(MultiPlan);
+		distributedPlan->planningError =
+			DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+						  "could not create distributed plan",
+						  "Possibly this is caused by the use of parameters in SQL "
+						  "functions, which is not supported in Citus.",
+						  "Consider using PLPGSQL functions instead.");
 	}
 
-	return physicalPlan;
+	/*
+	 * Error out if none of the planners resulted in a usable plan, unless the
+	 * error was possibly triggered by missing parameters.  In that case we'll
+	 * not error out here, but instead rely on postgres' custom plan logic.
+	 * Postgres re-plans prepared statements the first five executions
+	 * (i.e. it produces custom plans), after that the cost of a generic plan
+	 * is compared with the average custom plan cost.  We support otherwise
```

Comment:
or in more detail, that the only cost associated with a custom Citus plan is PostgreSQL's planning cost estimate.